File: java/cours/src/session3/TD/MotIterable/MotConcat.java
package session3.TD.MotIterable;

class MotConcat implements Mot {

	private Mot gauche;
	private Mot droit;
	private int taille;
	
	private char element;
	private Mot reste;

	public MotConcat(Mot g, Mot d) {
		this.gauche = g;
		this.droit = d;
		this.taille = g.taille() + d.taille();
	}

	@Override
	public int taille() {
		return this.taille;
	}

	@Override
	public boolean estConcat() {
		return true;
	}
	
	private void decomposer(){
		Mot courant = this;
		while(true){
			if(courant.estMotNonVide()){ // Cons
				this.reste = courant.reste();
				this.element = courant.element();
				return;
			}else{ // concat
				if(courant.gauche().estVide()){
					courant = courant.droit();
				}else if(courant.gauche().estMotNonVide()){
					this.reste = courant.gauche().reste().concat(courant.droit());
					this.element = courant.gauche().element();
					return;
				}else{
					courant = courant.gauche().gauche().concat(courant.gauche().droit().concat(courant.droit()));
				}
			}
		}
	}
	
	@Override
	public char element() {
		if(this.estVide()){
			throw new UnsupportedOperationException();
		}
		if(this.reste != null){
			return this.element;
		}
		decomposer();
		return this.element;
	}
	
	@Override
	public Mot reste() {
		if(this.estVide()){
			throw new UnsupportedOperationException();
		}
		if(this.reste != null){
			return this.reste;
		}
		decomposer();
		return this.reste;
	}

	@Override
	public Mot gauche() {
		return this.gauche;
	}

	@Override
	public Mot droit() {
		return this.droit;
	}

}

File: typescript/session3/MotIterable/MotConcat.ts
import { Mot } from "./mot";
import { Char } from "../Char";

export class MotConcat extends Mot {

  private motGauche: Mot;
  private motDroit: Mot;
  private tailleMot: number;

  private premiereLettre: Char;
  private resteMot: Mot;


  public constructor(g: Mot, d: Mot) {
    super();
    this.motGauche = g;
    this.motDroit = d;
    this.tailleMot = g.taille() + d.taille();
  }

  public taille(): number {
    return this.tailleMot;
  }

  public estConcat(): boolean {
    return true;
  }

  private decomposer(): void {
    var courant: Mot = this;
    while (true) {
      if (courant.estMotNonVide()) { // Cons
        this.resteMot = courant.reste();
        this.premiereLettre = courant.element();
        return;
      } else { // concat
        if (courant.gauche().estVide()) {
          courant = courant.droit();
        } else if (courant.gauche().estMotNonVide()) {
          this.resteMot = courant.gauche().reste().concat(courant.droit());
          this.premiereLettre = courant.gauche().element();
          return;
        } else {
          courant = courant.gauche().gauche().concat(courant.gauche().droit().concat(courant.droit()));
        }
      }
    }
  }

  public element(): Char {
    if (this.estVide()) {
      throw new Error("Unsupported Operation Exception");
    }
    if (this.premiereLettre != null) {
      return this.premiereLettre;
    }
    this.decomposer;
    return this.premiereLettre;
  }

  public gauche(): Mot {
    return this.motGauche;
  }
  public droit(): Mot {
    return this.motDroit;
  }
}